Refactor TimestampContainer.

Bopchy opened the pull request:
Blocked on #1619 

Resolves #1565 

**Overall change:** _Refactor TimestampContainer._

**Code changes:**

- _Create ArticleTimestamp container, and move the Article specific timestamp functionalities within TimestampContainer to it. This is so that the TimestampContainer can become a more generic container._
- _Add unit tests and storybook story for new ArticleTimestamp container. Storybook shows default output of ArticleTimestamp container._
- _Update TimestampContainer story to show variations of TimestampContainer  output._

**Links for testing:**
- Storybook Article Timestamp: http://localhost:9001/?selectedKind=ArticleTimestamp
- Storybook Timestamp: http://localhost:9001/?selectedKind=TimestampContainer

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [x] Test engineer approval
- [x] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

*NOTE:* Unit tests are still a WIP

Commit:
Create ArticleTimestamp. Refactor TimestampContainer.
- Create ArticleTimestamp on storybook.
- Edit TimestampContainer on storybook.

Commit:
Fix time not showing up correctly on storybook.

Commit:
Add ArticleTimestamp tests.

Commit:
Update TimestampContainer tests. Fix eslint issues.

Commit:
Move relativeTime logic to Timestmap container.

Commit:
Update tests. Fix linting issues.

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into 1565-refactor-timestamp-container-logic

Commit:
Add tests.
- Add Timestamp/ tests to increase coverage.

Commit:
Refactor test to remove repetition.

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into 1565-refactor-timestamp-container-logic

Commit:
Merge branch 'latest' into 1565-refactor-timestamp-container-logic

pjlee11 commented on src/app/containers/ArticleTimestamp/__snapshots__/index.test.jsx.snap:
```diff
@@ -0,0 +1,133 @@
+
+exports[`ArticleTimestamp should handle incomplete data 1`] = `null`;
+
+exports[`ArticleTimestamp should not return values for invalid timestamps 1`] = `
```
Would we expect this to render an empty `<div>` as a grid item? I'd assume if we aren't rendering a timestamp value we shouldn't render the `<div>`

Bopchy replied:
Added a check to stop empty `<div>` being rendered.

Commit:
Integrate one or two timestamps changes.

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into 1565-refactor-timestamp-container-logic

Commit:
Remove comments.

Commit:
Merge branch '1565-refactor-timestamp-container-logic' of github.com:bbc/simorgh into 1565-refactor-timestamp-container-logic

Commit:
Add test for invalid ArticleTimestamp data.

dr3 commented on src/app/containers/Timestamp/index.jsx:
```diff
@@ -1,86 +1,51 @@
-  lastPublished: number.isRequired,
+  timestamp: number.isRequired,
+  dateTimeFormat: string.isRequired,
+  format: string.isRequired,
```
Strictly `format` shouldn't be required, as if `isRelative` is true its not used

dr3 commented on src/app/containers/Timestamp/index.jsx:
```diff
@@ -1,86 +1,51 @@
+  format: string.isRequired,
+  isRelative: bool.isRequired,
+  timezone: string,
+  local: string,
```
I dont think `local` is used

Bopchy replied:
It isn't being used. Removed it.

dr3 commented on src/app/containers/Timestamp/index.jsx:
```diff
@@ -1,86 +1,51 @@
+  timestamp: number.isRequired,
+  dateTimeFormat: string.isRequired,
+  format: string.isRequired,
+  isRelative: bool.isRequired,
```
Also can `isRelative` be not required please? :) Personal preference but I would say default to false maybe?

sareh commented on src/app/containers/ArticleTimestamp/helpers.js:
```diff
@@ -0,0 +1,30 @@
+
+export const isToday = timestamp => isSameDay(Date.now(), timestamp);
+
+export const formatType = timestamps => {
```
Could you please change this function `formatType` function to take in params `({lastPublished, firstPublished})`. This can simplify the usage in the ArticleTimestamp file https://github.com/bbc/simorgh/pull/1617/files#diff-9112a95767a6e83c378bf24186177918R26 
and make it explicitly clear what we're checking here in this function. For example, that we're always checking `isToday(lastUpdated)`.

sareh commented on src/app/containers/ArticleTimestamp/index.jsx:
```diff
@@ -0,0 +1,40 @@
+import React from 'react';
```
Great work on the refactor. This Article Timestamp container is much more readable. ✨

sareh commented on src/app/containers/ArticleTimestamp/index.stories.jsx:
```diff
@@ -0,0 +1,31 @@
+});
+
+storiesOf('ArticleTimestamp', module)
+  .add('default', () => (
```
Can we please have all 8 scenarios for the Article Container as stories here? This will give us all visibility of the various formats that the Article Container can have. 

You can see the scenarios in the description here: https://github.com/bbc/simorgh/pull/1619 Or by just looking at the unit tests.

sareh replied:
@Bopchy Thanks for adding all of these scenarios. It's great to have visibility of the variants of this container in Storybook.

sareh commented on src/app/containers/Timestamp/index.jsx:
```diff
@@ -1,86 +1,51 @@
+};
+
+TimestampContainer.defaultProps = {
+  local: 'en-GB',
```
Can this prop name please be updated to be `locale`? This is to be in keeping with usage elsewhere.

Bopchy replied:
Ended up removing this prop because it wasn't being used.

sareh commented on src/app/containers/Timestamp/relativeTimestamp.js:
```diff
@@ -2,7 +2,10 @@ const constructTimestamp = timeDiffMins => {
   let unit;
 
-  if (timeDiffMins >= 60) {
+  if (timeDiffMins >= 1440) {
```
Is there a reason for adding support for days?

Bopchy replied:
Days support was added so that relativeTimestamp() could be generic enough to capture any time period e.g. 1 hour ago, 23 days ago

sareh reviewed (changes requested):
Great work on this PR, Ruth! The timestamps are much more understandable like this and also reusable. I've added a few comments below.

Commit:
Make requested changes.
- Introduce showRelativeTime function to handle time format shown.
- Make isRelative, format optional props.
- Remove local prop.
- Add ArticleTimestamp scenarios.
- Edit formatType().

Commit:
Merge remote-tracking branch 'origin' into 1565-refactor-timestamp-container-logic

dr3 commented on src/app/containers/Timestamp/timestampUtilities.js:
```diff
@@ -14,3 +15,13 @@ export const formatUnixTimestamp = (timestamp, momentString, timezone) =>
+  if (isRelative) {
+    return relativeTime(timestamp);
+  }
+  if (!isRelative && format) {
```
As `isRelative` is checked above and returned out of, it doesnt need to be checked here :)

dr3 commented on src/app/containers/Timestamp/timestampUtilities.js:
```diff
@@ -14,3 +15,13 @@ export const formatUnixTimestamp = (timestamp, momentString, timezone) =>
+  if (!isRelative && format) {
+    return formatUnixTimestamp(timestamp, format, timezone);
+  }
+  return formatUnixTimestamp(timestamp, 'D MMMM YYYY, HH:mm z', timezone);
```
Not sure if I like the default being set here. Curious of others thoughts, at the least it can be declared as `const defaultFormat` at the top of this function :P

Bopchy replied:
Declared a `defaultFormat` to hold the default.

pjlee11 commented on src/app/containers/Timestamp/__snapshots__/index.test.jsx.snap:
```diff
@@ -0,0 +1,107 @@
+  dateTime="2018-10-19"
+>
+  It was 
+  192 days ago
```
Are we certain this isn't going to change with the date? EG: tomorrow being "193 days ago"

Bopchy replied:
Changed the test so that the date used remains constant, and doesn't change with the date.

Commit:
Create default time format variable. Make snapshot test date constant.

sareh reviewed (approved):
👍 Great work on this PR, Ruth! Thanks for making all the suggested changes. 

Before putting the ticket in ready for testing, it's good to add test instructions onto the PR. In this case, since we have all of the scenarios in Storybook - linking to the stories would be useful http://localhost:9001/?selectedKind=ArticleTimestamp & http://localhost:9001/?selectedKind=TimestampContainer

david-boydell commented:
I know it's pretty noddy but could the `datetime` values have a day value over 12? That way you only have to glance at the datetime attr to know it's been properly formatted; values such as datetime="2018-07-07" are ambiguous.

jamesbrumpton commented:
LGTM. Happy for this to be merged.
